### PR TITLE
feat(world,store): add ERC165 checks for all registration methods

### PR DESCRIPTION
**Pull request by alvrs**

### CODE REVIEW

**Comment by alvrs** [comment]
Current dependencies on/for this PR:
* main
  * **PR #1457** <a href="https://app.graphite.dev/github/pr/latticexyz/mud/1457" target="_blank">Graphite</a> 
    * **PR #1458** <a href="https://app.graphite.dev/github/pr/latticexyz/mud/1458" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/latticexyz/mud/1458?utm_source=stack-comment).

File: packages/store/abi/ERC165.sol/ERC165.abi.json [hunk 1]
```diff
@@ -0,0 +1,21 @@
+[
```
Comment (holic) [inline packages/store/abi/ERC165.sol/ERC165.abi.json:1]: I added `.gitattributes` in another PR but might be worth adding here to land it sooner to suppress these codegen diffs for now:
https://github.com/latticexyz/mud/pull/1354/files#diff-618cd5b83d62060ba3d027e314a21ceaf75d36067ff820db126642944145393e

File: packages/world/src/interfaces/ISystemHook.sol [hunk 2]
```diff
@@ -1,7 +1,13 @@
 // SPDX-License-Identifier: MIT
 pragma solidity >=0.8.0;
 
-interface ISystemHook {
+import { ERC165 } from "./ERC165.sol";
+
+// ERC-165 Interface ID (see https://eips.ethereum.org/EIPS/eip-165)
+bytes4 constant SYSTEM_HOOK_INTERFACE_ID = ISystemHook.onBeforeCallSystem.selector ^
+  ISystemHook.onAfterCallSystem.selector;
```
Comment (holic) [inline packages/world/src/interfaces/ISystemHook.sol:8]: I noticed `IModule` above includes `ERC165.supportsInterface.selector` but this does not. Any reason?

File: packages/store/src/IStoreHook.sol [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+import { Schema } from "./Schema.sol";
+import { ERC165 } from "./ERC165.sol";
+
+// ERC-165 Interface ID (see https://eips.ethereum.org/EIPS/eip-165)
+bytes4 constant STORE_HOOK_INTERFACE_ID = IStoreHook.onBeforeSetRecord.selector ^
+  IStoreHook.onAfterSetRecord.selector ^
+  IStoreHook.onBeforeSetField.selector ^
+  IStoreHook.onAfterSetField.selector ^
+  IStoreHook.onBeforeDeleteRecord.selector ^
+  IStoreHook.onAfterDeleteRecord.selector;
```
Comment (holic) [inline packages/store/src/IStoreHook.sol:13]: same here, seems to be missing `ERC165.supportsInterface.selector`?

**Review by holic — APPROVED** [review]

File: packages/world/src/interfaces/ISystemHook.sol (hunk 2 above)
Comment (alvrs) [reply packages/world/src/interfaces/ISystemHook.sol]: good catch, should be included

File: packages/world/src/interfaces/IDelegationControl.sol [hunk 4]
```diff
@@ -1,6 +1,16 @@
...
-interface IDelegationControl {
+import { IWorldContextConsumer } from "./IWorldContextConsumer.sol";
+import { ERC165 } from "./ERC165.sol";
+
+// ERC-165 Interface ID (see https://eips.ethereum.org/EIPS/eip-165)
+bytes4 constant DELEGATION_CONTROL_INTERFACE_ID = IDelegationControl.verify.selector ^
+  IWorldContextConsumer._msgSender.selector ^
+  IWorldContextConsumer._msgValue.selector ^
+  IWorldContextConsumer._world.selector ^
+  ERC165.supportsInterface.selector;
```
Comment (holic) [inline packages/world/src/interfaces/IDelegationControl.sol:12]: is it necessary to define the inherited selectors as well as our own?

and if so would it be easier to do the whole interface ID instead of each selector (caveat: I don't know what the bitshifting is doing here)?

```solidity
bytes4 constant DELEGATION_CONTROL_INTERFACE_ID = IDelegationControl.verify.selector ^
  WORLD_CONTEXT_CONSUMER_INTERFACE_ID ^
  ERC165_INTERFACE_ID;
```

File: packages/world/src/interfaces/IWorldContextConsumer.sol [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+// ERC-165 Interface ID (see https://eips.ethereum.org/EIPS/eip-165)
+bytes4 constant WORLD_CONTEXT_CONSUMER_INTERFACE_ID = IWorldContextConsumer._msgSender.selector ^
+  IWorldContextConsumer._msgValue.selector ^
+  IWorldContextConsumer._world.selector ^
+  ERC165.supportsInterface.selector;
+
+interface IWorldContextConsumer is ERC165 {
+  function _msgSender() external view returns (address);
+  function _msgValue() external view returns (uint256);
+  function _world() external view returns (address);
```
Comment (dk1a) [inline packages/world/src/interfaces/IWorldContextConsumer.sol]: why are these needed?

**Review by dk1a — APPROVED** [review]

File: packages/world/src/interfaces/IDelegationControl.sol (hunk 4 above)
Comment (alvrs) [reply packages/world/src/interfaces/IDelegationControl.sol:12]: agree this is cleaner, and should work since `^` (bitwise XOR) is commutative and associative

File: packages/world/src/interfaces/IWorldContextConsumer.sol (hunk 5 above)
Comment (alvrs) [reply packages/world/src/interfaces/IWorldContextConsumer.sol]: need them to be public to be part of the interface to be able to distinguish a `WorldContextConsumer` contract

File: packages/store/src/ERC165.sol [hunk 6]
```diff
@@ -0,0 +1,15 @@
+// SPDX-License-Identifier: MIT
+pragma solidity >=0.8.0;
+
+bytes4 constant ERC165_INTERFACE_ID = ERC165.supportsInterface.selector;
+
+// See https://eips.ethereum.org/EIPS/eip-165
+interface ERC165 {
```
[the remark this answers is not in the document]
Comment (alvrs) [reply packages/store/src/ERC165.sol:7]: agree

**Comment by alvrs** [comment]
merging since there were no significant changes since the last approval, just implementation of review feedback and rebasing
